### PR TITLE
Docs: Flow control section

### PR DESCRIPTION
**Pull request by djfarrelly**
Create a new section of the docs focused on all flow control articles. Includes an overview page and all pages extracted from the TypeScript docs into the guides section.

* [x] Flow control overview page
* [x] Rate limit
* [x] Debounce
* [x] Priority
* [x] Clean up TS reference docs

### CODE REVIEW

File: pages/docs/guides/flow-control.mdx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+Avoid unnecessary function invocations by de-duplicating events over a sliding time window. Ideal for preventing wasted work when a function might be triggered in quick succession.
+
+### <a href="/docs/guides/priority" className="flex items-center gap-4"><IconPriority size="2rem"/> Priority</a> {{anchor: false}}
+
+Dynamically adjust the execution order of functions based on any data. Ideal for pushing critical work to the front of the queue.
+
+### <a href="/docs/guides/handling-idempotency" className="flex items-center gap-4"><IconIdempotency size="2rem"/> Idempotency</a> {{anchor: false}}
+
+Prevent duplicate work with two different approaches to ensure functions are only run once.
```
Comment (sylwiavargas) [inline pages/docs/guides/flow-control.mdx:40]:
```suggestion
Prevent duplicate work with two different approaches to ensure functions run only once.
```

just Plain English -- unless it's important to emphasize that they are being run.

File: pages/docs/guides/debounce.mdx [hunk 2]
```diff
@@ -0,0 +1,72 @@
+export const description = 'Avoid unnecessary function invocations by de-duplicating events over a sliding time window. Ideal for preventing wasted work when a function might be triggered in quick succession.'
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:1]:
```suggestion
export const description = 'De-duplicate events over Avoid unnecessary function invocations by de-duplicating events over a sliding time window. Ideal for preventing wasted work when a function might be triggered in quick succession.'
```

File: pages/docs/guides/debounce.mdx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+# Debounce
+
+Debounce allows a function to delay execution until a series of events have stopped being received. This is useful for preventing wasted work when a function might be triggered in quick succession. Use cases for debounce include:
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:5]:
```suggestion
Debounce delays function execution until a series of events are no longer received. This is useful for preventing wasted work when a function might be triggered in quick succession. Use cases for debounce include:
```

File: pages/docs/guides/debounce.mdx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+* Prevent wasted work when handling events from user input that may change multiple times in a short time period
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:7]:
```suggestion
* Preventing wasted work when handling events from user input that may change multiple times in a short time period.
```

File: pages/docs/guides/debounce.mdx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+* Delay processing of noisy webhook events until they have stopped being received
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:8]:
```suggestion
* Delaying processing of noisy webhook events until they are no longer received.
```

File: pages/docs/guides/debounce.mdx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+* Ensuring that functions use the latest event within a series of updates (example: synchronization)
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:9]:
```suggestion
* Ensuring that functions use the latest event within a series of updates (for example, synchronization).
```

File: pages/docs/guides/debounce.mdx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  {
+    id: "handle-webhook",
+    debounce: {
+      key: "event.data.account_id",
+      period: "5m",
+    },
+  },
+  { event: "intercom/company.updated" },
+  async ({ event, step }) => {
+    // This function will only be scheduled 5m after events have stopped being received with the same
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:25]:
```suggestion
    // This function will only be scheduled 5 minutes after events are no longer received with the same
```

File: pages/docs/guides/debounce.mdx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    // `event` will be the last event in the series received.
+  }
+);
+```
+</CodeGroup>
+
+### Configuration reference
+
+* `period` - The time delay to delay execution. The period begins when the first matching event is received.
+* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has it's own limit, enabling you to rate limit function runs by any particular key, like a user id.
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:37]:
```suggestion
* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has its own limit, enabling you to rate limit function runs by any particular key, like a user ID.
```

File: pages/docs/guides/debounce.mdx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  alt="Visualization of how debounce is applied"
+/>
+
+### Using a `key`
+
+When a `key` is added, a separate debounce period is applied for each unique value of the `key` expression. For example if your `key` is set to `event.data.customer_id`, each customer would have their individual debounce period applied to functions run. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more info.
+
+## Comparison to rate limiting
+
+For use cases where you prefer to execute a function for the _first_ event received, consider using [rate limiting](/docs/guides/rate-limiting) instead. Rate limiting ensures that a function runs once for each `key` the *first* time an event is received, while debounce uses the *last* event during a specified period.
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:57]:
```suggestion
If you prefer to execute a function for the _first_ event received, consider using [rate limiting](/docs/guides/rate-limiting) instead. Rate limiting ensures that a function runs once for each `key` the *first* time an event is received, while debounce uses the *last* event during a specified period.
```

File: pages/docs/guides/debounce.mdx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+## Combining with idempotency
+
+Debounce can be combined with [idempotency](/docs/guides/handling-idempotency#at-the-function-level-the-consumer) to ensure that once a function is debounced it does not run again.
```
Comment (sylwiavargas) [inline pages/docs/guides/debounce.mdx:61]:
```suggestion
Debounce can be combined with [idempotency](/docs/guides/handling-idempotency#at-the-function-level-the-consumer) to ensure that once the debounced function has run, it does not run again.
```

File: pages/docs/guides/priority.mdx [hunk 11]
```diff
@@ -0,0 +1,84 @@
+export const description = 'Dynamically adjust the execution order of functions based on any data. Ideal for pushing critical work to the front of the queue.'
+
+# Priority
+
+Priority allows you to dynamically execute some runs ahead or behind others based on any data. This allows you to prioritize some jobs ahead others without the need for a separate queue. Some use cases for priority include:
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:5]:
```suggestion
Priority allows you to dynamically execute some runs ahead or behind others based on any data. This allows you to prioritize some jobs ahead of others without the need for a separate queue. Some use cases for priority include:
```

File: pages/docs/guides/priority.mdx [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+* Give higher priority based on a user's subscription level, for example, free vs. paid users
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:7]:
```suggestion
* Giving higher priority based on a user's subscription level, for example, free vs. paid users.
```

File: pages/docs/guides/priority.mdx [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+* Ensure that critical work is executed before other work in the queue
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:8]:
```suggestion
* Ensuring that critical work is executed before other work in the queue.
```

File: pages/docs/guides/priority.mdx [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+* Prioritize certain jobs for users during onboarding to give the user a better first run experience
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:9]:
```suggestion
* Prioritizing certain jobs during onboarding to give the user a better first-run experience.
```

File: pages/docs/guides/priority.mdx [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+## How priority works
+
+Functions are scheduled in a priority queue based on the time they should run. By default, all functions are enqueued at the current time (a factor of `0`). If a function has `priority` configured, Inngest evaluates the `run` expression for each new function run based on the input event's data. The `run` expression should return a factor, in seconds, (positive or negative) to adjust the priority of the function run.
+
+Expressions that return a **positive** number will **increase the priority** of the function run ahead of other jobs already in the queue by the given value in seconds. The function will be run ahead of other jobs that were enqueued up to that many seconds ago. For example, if a function run is scheduled with a factor of `120`, it will run ahead of any jobs enqueued in the last 120 seconds, given that they are still in the queue and have not completed.
+
+Expressions that return a **negative** number will **delay** the function run by the given value in seconds.
+
+**Practical example**
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:45]:
```suggestion
### Practical example
```

File: pages/docs/guides/priority.mdx [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+Given we have three jobs in the queue, each which was enqueued at the following times:
+
+```plaintext
+Jobs:          [A,        B,        C       ]
```
Comment (sylwiavargas) [inline pages/docs/guides/priority.mdx:50]: I like this very much, thanks

File: pages/docs/guides/rate-limiting.mdx [hunk 17]
```diff
@@ -0,0 +1,103 @@
+export const description = 'Prevent excessive function runs over a given time period by skipping events beyond a specific limit. Ideal for protecting against abuse.'
+
+# Rate limiting
+
+Rate limiting is a _hard limit_ on how many function runs can start within a time period. Events that exceed the rate limit are _skipped_ and do not trigger functions to start. This prevents excessive function runs over a given time period. Some use cases for rate limiting include:
+
+* Preventing abuse of your system
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:7]:
```suggestion
* Preventing abuse of your system.
```

File: pages/docs/guides/rate-limiting.mdx [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+* Reducing frequency of data synchronization functions
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:8]:
```suggestion
* Reducing frequency of data synchronization functions.
```

File: pages/docs/guides/rate-limiting.mdx [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+* Smoothing out noisy [webhook](/docs/platform/webhooks) events
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:9]: Not sure what "Smoothing out" means here. Would you have a common or Plain English expression to offer here?

File: pages/docs/guides/rate-limiting.mdx [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+    rateLimit: {
+      limit: 1,
+      period: "4h",
+      key: "event.data.company_id",
+    },
+  },
+  { event: "intercom/company.updated" },
+  async ({ event, step }) => {
+    // This function will be rate limited
+    // It will only run 1 once per 4 hours for a given event payload with matching company_id
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:27]:
```suggestion
    // It will only run once per 4 hours for a given event payload with matching company_id
```

File: pages/docs/guides/rate-limiting.mdx [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+  }
+);
+```
+</CodeGroup>
+
+### Configuration reference
+
+* `limit` - The maximum number of functions to run in the given time period.
+* `period` - The time period of which to set the limit. The period begins when the first matching event is received.
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:36]: "The time period of which to set the limit." doesn't sound intuitive to me but I don't have a suggestion atm on how to make it easy.

File: pages/docs/guides/rate-limiting.mdx [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has it's own limit, enabling you to rate limit function runs by any particular key, like a user id.
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:37]:
```suggestion
* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has its own limit, enabling you to rate limit function runs by any particular key, like a user id.
```

File: pages/docs/guides/rate-limiting.mdx [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+<Callout>
+  Any events received in excess of your `limit` are ignored. This means this is not the right approach if you need to process every single event sent to Inngest. Instead, check out [throttle](/docs/guides/throttling).
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:40]:
```suggestion
  Any events received in excess of your `limit` are ignored. This means this is not the right approach if you need to process every single event sent to Inngest. Consider using [throttle](/docs/guides/throttling) instead.
```

File: pages/docs/guides/rate-limiting.mdx [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+</Callout>
+
+## How rate limiting works
+
+Each time an event is received that match's your function's trigger, it is evaluated prior to executing your function. If `rateLimit` is configured, Inngest uses the `limit` and `period` options to only execute a maximum number of functions during that period.
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:45]:
```suggestion
Each time an event is received that matches your function's trigger, it is evaluated prior to executing your function. If `rateLimit` is configured, Inngest uses the `limit` and `period` options to only execute a maximum number of functions during that period.
```

File: pages/docs/guides/rate-limiting.mdx [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+limit / period = bucket time window
+```
+
+For example, this means that for a `limit: 10` and `period: '60m'` (60 minutes), the bucket time window will be 6 minutes. Any event triggering the function "fills up" the bucket for that time window and any additional events are ignored until the bucket's time window is reset. The algorithm (GCRA) is more sophisticated than this, but at the basic level - `rateLimit` ensures that you'll only run the max `limit` number of items over the `period` that you specify.
+
+<Callout>
+  Events that are ignored by the function will continue to be stored by Inngest.
+</Callout>
+
+**How the rate limit is applied with a consistent rate of events received**
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:59]: I love these graphics, thank you for creating them!

Two comments, though:
1. To some folks diagrams are not intuitive. Could you add a description of what is seen?
2. Less importantly, could you check if the reversed colors work here? If yes, we could easily add a dark-mode version (I have created a component like this a while ago, used it in Resend's guide I think)

File: pages/docs/guides/rate-limiting.mdx [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+  src="/assets/docs/reference/function-rate-limit-equal-1.png"
+  width="800"
+  height={280}
+  quality="95"
+  alt="Visualization of how the rate limit is applied when limit is set to 1"
+/>
+
+### Using a `key`
+
+When a `key` is added, a separate limit is applied for each unique value of the `key` expression. For example if your `key` is set to `event.data.customer_id`, each customer would have their individual rate limit applied to functions run meaning different users might have the same function run in same bucket time window, but two runs will not happen for the same `event.data.customer_id`. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more info.
```
Comment (sylwiavargas) [inline pages/docs/guides/rate-limiting.mdx:91]:
```suggestion
When a `key` is added, a separate limit is applied for each unique value of the `key` expression. For example, if your `key` is set to `event.data.customer_id`, each customer would have their individual rate limit applied to functions run meaning different users might have the same function run in same bucket time window, but two runs will not happen for the same `event.data.customer_id`. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more information.
```

File: pages/docs/guides/throttling.mdx [hunk 27]
```diff
@@ -1,10 +1,16 @@
+export const description = 'Limit the throughput of function execution over a period of time. Ideal for working around third-party API rate limits.';
+
 # Throttling
 
-Throttling allows you to specify how many function runs can start within a time period.  When the limit is reached, new function runs over the throttling limit will be enqueued for the future.  This smooths how often functions start running in a given time period.
+Throttling allows you to specify how many function runs can start within a time period. When the limit is reached, new function runs over the throttling limit will be _enqueued for the future_. Some use cases for priority include:
+
+* Smooth out execution of functions over time
```
Comment (sylwiavargas) [inline pages/docs/guides/throttling.mdx:7]:
```suggestion
* Smoothing out execution of functions over time.
```

I'm not sure what that means. Could you rephrase "Smoothing out"?

File: pages/docs/guides/throttling.mdx [hunk 28]
```diff
@@ -1,10 +1,16 @@
...
+* Working around third-party API rate limits
```
Comment (sylwiavargas) [inline pages/docs/guides/throttling.mdx:8]:
```suggestion
* Working around third-party API rate limits.
```

File: pages/docs/guides/throttling.mdx [hunk 29]
```diff
@@ -20,15 +26,16 @@ inngest.createFunction(
...
+</CodeGroup>
 
 You can configure throttling on each function using the optional `throttle` parameter.  The options directly control the generic cell rate algorithm parameters used within the queue.
 
 ### Configuration reference
 
 - `limit`: The total number of runs allowed to start within the given `period`.
-- `period`: The period for the rate limit.
+- `period`: The period within the limit will be applied.
 - `burst`: The number of runs allow to start in the given window in a single burst. This defaults to 1, which ensure that requests are smoothed amongst the given `period`.
```
Comment (sylwiavargas) [inline pages/docs/guides/throttling.mdx:37]:
```suggestion
- `burst`: The number of runs allowed to start in the given window in a single burst. This defaults to 1, which ensures that requests are smoothed amongst the given `period`.
```

File: pages/docs/reference/functions/create.mdx [hunk 30]
```diff
@@ -43,6 +43,25 @@ The `createFunction` method accepts a series of arguments to define your functio
...
+
+    <Properties nested={true} collapse={true}>
+      <Property name="limit" type="number" required>
+        The total number of runs allowed to start within the given `period`.
+      </Property>
+      <Property name="period" type="number" required>
+        The period within the `limit` will be applied.
+      </Property>
+      <Property name="burst" type="number">
+        The number of runs allow to start in the given window in a single burst. This defaults to 1, which ensure that requests are smoothed amongst the given `period`.
```
Comment (sylwiavargas) [inline pages/docs/reference/functions/create.mdx:57]:
```suggestion
        The number of runs allowed to start in the given window in a single burst. This defaults to 1, which ensures that requests are smoothed amongst the given `period`.
```

File: pages/docs/reference/functions/run-priority.mdx [hunk 31]
```diff
@@ -1,94 +1,27 @@
 # Function run priority <VersionBadge version="v3.2.1+" />
 
-Defining a function's run priority allows you to **prioritize specific function runs above other runs within the same function**.  For example:
+Defining a function's run priority allows you to **prioritize specific function runs above other runs within the same function**.
```
Comment (sylwiavargas) [inline pages/docs/reference/functions/run-priority.mdx:3]:
```suggestion
You can prioritize specific function runs above other runs **within the same function**.
```

**Review by sylwiavargas — APPROVED** [review]
FANTASTIC SECTION. Thank you for writing it, and for making sure that the pages follow a similar structure. I also appreciated how you cross-linked them. Thank you.

I left a few edits and comments but approving it already as this PR is gold 🏆

File: pages/docs/guides/debounce.mdx (hunk 7 above)
Comment (djfarrelly) [reply pages/docs/guides/debounce.mdx:25]: Committing this, but it still feels weird - may be something to adjust later 🤔

File: pages/docs/guides/priority.mdx (hunk 16 above)
Comment (djfarrelly) [reply pages/docs/guides/priority.mdx:50]: We had something like this before, but I re-worked it a lot b/c it previously was pretty abstract, I used more clear timestamps to make it more obvious. I think this may be something that we could use a graphic for as well to make it clearer.

File: pages/docs/guides/priority.mdx [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+- Job Y: factor 120
+```
+
+Then Job Y will run ahead of Job X. Job Y will also run before any jobs scheduled 120 seconds beforehand. The queue will look like this:
+
+```plaintext
+Jobs:          [A,        Y,        B,        C,        X     ]
+Priority/Time: [12:00:10, 12:00:30, 12:00:40, 12:02:10, 12:02:30]
+                          │                             │
+                          └ 12:02:30 - 120s = 12:00:30   └ 12:02:30 - 0s = 12:02:30
```
Comment (djfarrelly) [inline pages/docs/guides/priority.mdx:67]:
```suggestion
                          └ 12:02:30 - 120s = 12:00:30  └ 12:02:30 - 0s = 12:02:30
```

File: pages/docs/guides/rate-limiting.mdx (hunk 25 above)
Comment (djfarrelly) [reply pages/docs/guides/rate-limiting.mdx:59]: Good points! I think this is worth while, just may not be able to get it done properly for release this morning 😄
